packages/graphql/src/registry.ts

```diff
@@ -0,0 +1,46 @@
+import { navigateProgram, type Program, type SemanticNodeListener } from "@typespec/compiler";
+import type { GraphQLObjectType } from "graphql";
+
+type Mutable<T> = {
+  -readonly [k in keyof T]: T[k];
+};
+
+// This class contains the registry of all the GraphQL types that are being used in the program
+export class GraphQLTypeRegistry {
+  program: Program;
+  readonly programNavigated: boolean = false;
+
+  constructor(program: Program) {
+    this.program = program;
+    return new Proxy(this, {
+      get(target: GraphQLTypeRegistry, prop: string, receiver) {
+        if (GraphQLTypeRegistry.#publicGetters.includes(prop)) {
+          if (!target.programNavigated) {
+            const mutableThis = target as Mutable<GraphQLTypeRegistry>;
+            navigateProgram(target.program, target.#semanticNodeListener);
+            mutableThis.programNavigated = true;
+          }
+        }
+        return Reflect.get(target, prop, receiver);
+      },
+    });
+  }
+
+  static get #publicGetters() {
+    return Object.entries(Object.getOwnPropertyDescriptors(GraphQLTypeRegistry.prototype))
+      .filter(([key, descriptor]) => {
+        return typeof descriptor.get === "function" && key !== "constructor";
+      })
+      .map(([key]) => key);
+  }
+
+  get rootQueryType(): GraphQLObjectType | undefined {
```

Comment:
Why do we need to make this dynamic/lazy? We can use `stateMaps` to handle state if we want (although we should check that with Microsoft and I'm not sure if that'll work for all our cases, but that is a different discussion). The navigateProgram already takes care of navigating the program, so I don't think we need a dynamically accessible get function.

I think a more readable and deterministic navigate flow looks like

navigateProgram --> enterXXX extract information based on type and add to some map --> exitXXX extract information from map and create a GraphQL AST object.

None of the existing emitters need lazy initialization, so I am not sure that the lazy init is doing much for us here.

Reply:
The getters in the prototype are only used in the exit method, so these shouldn't really be public. Maybe registry is not the right name, but this is the thing that should return a schema from a public method.

Reply:
Great! I was hoping to have this discussion 🙂 

The only thing that's intended to be lazy in `GraphQLTypeRegistry` is the execution of `navigateProgram`. Basically, we have this object/interface that we want to be able to provide various information about the schema, but in order to have that information we need to have walked through the TSP program via `navigateProgram`. The intent of this `Proxy` trap is to ensure that, whenever some of that data is accessed, we have already navigated the program and therefore that data is available.

The emitter flow then looks something like this:
1. `GraphQLEmitter` knows how to construct a GraphQL schema — i.e. it knows that you start with 1-3 root object types, each of those object types have fields, each of those fields are of various types, etc.
2. As `GraphQLEmitter` works on constructing the schema, it asks `GraphQLTypeRegistry` various GraphQL-y questions like: "is there a root Query type?" "What fields are on this object type?" "What does this decorator look like?"
3. `GraphQLTypeRegistry` interprets the TypeSpec schema in order to provide answers to those questions. It does so by walking through the TSP program and storing the answers to the questions that it is capable of asking.

This differs from the flow you outlined in that there is no explicit "now we must navigate the program" step. Instead of it being a TypeSpec-centric approach, it is a GraphQL-centric approach. This matters when we are looking to emit multiple GraphQL schemas — we want to start with each schema that we are going to emit and figuring out what it looks like, rather than starting with the whole TypeSpec program and figuring out what schemas we need to create, then categorizing all of the data into those different schemas (which requires a lot of global state). We already know what schemas we need to emit from the `@schema` decorator's stateMap.

From my understanding of the TypeSpec code, this is more in line with how both the existing `TypeEmitter` structure and the experimental `TypeKit` approaches work. They focus on creating definitions in the emitted language, and provide a series of questions that can be asked about the TypeSpec program in order to do so.

Also — I fully intend/expect that we will use `stateMap` for handling state where it can do so.